Qcut sampling postprocessing function

anthayes92 opened the pull request:
**Context:**
The pipeline for cutting a circuit with sampling measurements requires it own postprocessing capabilities. This will support returning bitstrings as well as expectation values, the choice of which should be specified by the user. 

**Description of the Change:**
Add sampling postprocessing function

**Benefits:**
Facilitates cutting circuits with sampling measurements

**Possible Drawbacks:**
Potential computational bottleneck depending on number of fragments circuits to be processed

anthayes92 commented:
[sc-15890]

trbromley commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,79 @@ def test_mc_state_prep(self):
+    gives the correct results.
+    """
+
+    def test_sample_postprocess(self):
```
Might be good to have an interface-based test like [here](https://github.com/PennyLaneAI/pennylane/blob/482cd0894548adeb1d507d7debf84aa321cc0726/tests/transforms/test_qcut.py#L2212), where we also check that the returned is converted to the right interface type.

trbromley commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,79 @@ def test_mc_state_prep(self):
+
+        assert np.isclose(postprocessed, expected)
+
+
```
Perhaps worth adding a quick test for `_reshape_and_find_degrees`.

anthayes92 replied:
Added thanks!

trbromley reviewed (changes requested):
Thanks @anthayes92!

angusjlowe commented on pennylane/transforms/qcut.py:
```diff
@@ -827,6 +827,116 @@ def expand_fragment_tapes_mc(
+    out_degrees = [d for _, d in communication_graph.out_degree]
+
+    evals = (0.5, 0.5, 0.5, -0.5, 0.5, -0.5, 0.5, -0.5)
+    expvals = []
+    for result, setting in zip(results, settings.T):
+        sample_terminal = []
+        sample_mid = []
+
+        for fragment_result, out_degree in zip(result, out_degrees):
+            sample_terminal.append(fragment_result[: -out_degree or None])
+            sample_mid.append(fragment_result[-out_degree or len(fragment_result) :])
+
+        sample_terminal = np.hstack(sample_terminal)
+        sample_mid = np.hstack(sample_mid)
+
+        assert set(sample_terminal).issubset({np.array(0), np.array(1)})
+        assert set(sample_mid).issubset({np.array(-1), np.array(1)})
+        # following Eq.(35) of Peng et.al: https://arxiv.org/abs/1904.00102
+        f = classical_processing_fn(sample_terminal)
+        sigma_s = np.prod(sample_mid)
+        t_s = f * sigma_s
+        c_s = np.prod([evals[s] for s in setting])
+        K = len(sample_mid)
+        expvals.append(8**K * c_s * t_s)
```
```suggestion
    vals = []
    for result, setting in zip(results, settings.T):
        sample_terminal = []
        sample_mid = []

        for fragment_result, out_degree in zip(result, out_degrees):
            sample_terminal.append(fragment_result[: -out_degree or None])
            sample_mid.append(fragment_result[-out_degree or len(fragment_result) :])

        sample_terminal = np.hstack(sample_terminal)
        sample_mid = np.hstack(sample_mid)

        assert set(sample_terminal).issubset({np.array(0), np.array(1)})
        assert set(sample_mid).issubset({np.array(-1), np.array(1)})
        # following Eq.(35) of Peng et.al: https://arxiv.org/abs/1904.00102
        f = classical_processing_fn(sample_terminal)
        sigma_s = np.prod(sample_mid)
        t_s = f * sigma_s
        c_s = np.prod([evals[s] for s in setting])
        K = len(sample_mid)
        vals.append(8**K * c_s * t_s)
```

anthayes92 commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,157 @@ def test_mc_state_prep(self):
+
+        fixed_settings = np.array([[0, 7, 1], [5, 7, 2], [1, 0, 3], [5, 1, 1]])
+
+        # spy_func = mocker.spy(fn)
```
I'm not sure how to spy on this function, generally we use `mocker.spy(package, "function")` but this function exists by itself so I only have a single arg to pass to spy

trbromley replied:
Ok, can leave out for now :+1:

trbromley commented on pennylane/transforms/qcut.py:
```diff
@@ -827,6 +827,118 @@ def expand_fragment_tapes_mc(
+
+def qcut_processing_fn_sample(
+    results: Sequence, communication_graph: MultiDiGraph, shots: int
+) -> List[np.array]:
```
```suggestion
) -> List:
```
This could be e.g. a TF tensor

trbromley commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,157 @@ def test_mc_state_prep(self):
+        ]
+
+        postprocessed = qcut.qcut_processing_fn_sample(fixed_samples, communication_graph, shots)
+        postprocessed = qml.math.cast_like(postprocessed, lib.ones(1))
```
Why are we casting the output to the interface type? We need to cast the input to the interface type and check that the output is of the same type.

trbromley replied:
(same for the test below)

trbromley commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,157 @@ def test_mc_state_prep(self):
+
+        reshaped = qcut._reshape_results(results, communication_graph, shots)
+
+        for resh, exp_resh in zip(reshaped, expected_reshaped):
+            for arr, exp_arr in zip(resh, exp_resh):
+                assert np.allclose(arr, exp_arr)
```
Does a big `np.allclose(reshaped, expected_reshaped)` not work because the arrays contain objects?

anthayes92 replied:
exactly, also why there is a second for loop

trbromley reviewed (commented):
Thanks @anthayes92! Just left a few more comments - would it also be possible to fix what is mentioned [here](https://github.com/PennyLaneAI/pennylane/pull/2358#discussion_r833538455)?

anthayes92 commented on pennylane/transforms/qcut.py:
```diff
@@ -827,6 +827,118 @@ def expand_fragment_tapes_mc(
+        sample_terminal = np.hstack(sample_terminal)
+        sample_mid = np.hstack(sample_mid)
+
+        assert set(sample_terminal).issubset({np.array(0), np.array(1)})
```
yes that's what I found!

anthayes92 commented:
Thanks for the suggestions @trbromley , I've updated the docsting for `expand_fragment_tapes_mc` to include the `settings` array and also changed the tape diagrams since the `tape.draw()` function has been updated to give cleaner output.

trbromley commented on pennylane/transforms/qcut.py:
```diff
@@ -827,6 +824,118 @@ def expand_fragment_tapes_mc(
+        Check out the :func:`qml.cut_circuit_mc() <pennylane.cut_circuit_mc>` transform for more details.
+
+    Args:
+        results (Sequence): a collection of execution results generated from the
```
```suggestion
        results (Sequence): a collection of sample-based execution results generated from the
```

trbromley commented on pennylane/transforms/qcut.py:
```diff
@@ -827,6 +824,118 @@ def expand_fragment_tapes_mc(
+        Check out the :func:`qml.cut_circuit_mc() <pennylane.cut_circuit_mc>` transform for more details.
+
+    Args:
+        results (Sequence): a collection of execution results generated from the
```
```suggestion
        results (Sequence): a collection of sample-based execution results generated from the
```

trbromley commented on pennylane/transforms/qcut.py:
```diff
@@ -827,6 +824,118 @@ def expand_fragment_tapes_mc(
+            random expansion of circuit fragments over measurement and preparation node configurations
+        communication_graph (nx.MultiDiGraph): the communication graph determining connectivity
+            between circuit fragments
+        settings (tensor_like): each element is one of 8 unique values that determines and tracks the specific
```
```suggestion
        settings (np.ndarray): Each element is one of 8 unique values that tracks the specific
```
We have control over the type of `settings`.

trbromley commented on pennylane/transforms/qcut.py:
```diff
@@ -827,6 +824,118 @@ def expand_fragment_tapes_mc(
+        shots (int): the number of shots
+        classical_processing_fn (callable): a classical postprocessing function to be applied to
+            the reconstructed bitstrings. The expected input is a bitstring; a flat array of length ``wires``
+            and the output should be a single number within the interval [-1, 1]
```
```suggestion
            and the output should be a single number within the interval :math:`[-1, 1]`.
```

trbromley commented on pennylane/transforms/qcut.py:
```diff
@@ -827,6 +824,118 @@ def expand_fragment_tapes_mc(
+            measurement and preparation operations over all configurations. The number of rows is determined
+            by the number of cuts and the number of columns is determined by the number of shots.
+        shots (int): the number of shots
+        classical_processing_fn (callable): a classical postprocessing function to be applied to
```
```suggestion
        classical_processing_fn (callable): A classical postprocessing function to be applied to
```

trbromley commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
+
+        communication_graph = MultiDiGraph(frag_edge_data)
+        shots = 3
+        fragment_configurations, settings = qml.transforms.qcut.expand_fragment_tapes_mc(
+            fragment_tapes, communication_graph, shots
+        )
```
```suggestion
```
Do we need this?

trbromley replied:
BTW, not sure what IDE you are using, but on PyCharm this is easy to see with unused variables shown in a faded grey colour:
![image](https://user-images.githubusercontent.com/49409390/160293291-e193de30-33d0-4073-9d79-4ba711f4f07c.png)

Maybe something that can be enabled?

trbromley commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
+            np.array([[0.0], [-1.0], [-1.0]]),
+            np.array([[1.0], [1.0], [1.0]]),
+        ]
+        cast_fixed_samples = [qml.math.cast_like(fs, lib.ones(1)) for fs in fixed_samples]
```
```suggestion
        cast_fixed_samples = [qml.math.convert_like(fs, lib.ones(1)) for fs in fixed_samples]
```
This is an important distinction.

trbromley commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
+
+        communication_graph = MultiDiGraph(frag_edge_data)
+        shots = 3
+        fragment_configurations, settings = qml.transforms.qcut.expand_fragment_tapes_mc(
+            fragment_tapes, communication_graph, shots
+        )
```
```suggestion
```

trbromley commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
+            np.array([[0.0], [-1.0], [-1.0]]),
+            np.array([[1.0], [1.0], [1.0]]),
+        ]
+        cast_fixed_samples = [qml.math.cast_like(fs, lib.ones(1)) for fs in fixed_samples]
```
```suggestion
        cast_fixed_samples = [qml.math.convert_like(fs, lib.ones(1)) for fs in fixed_samples]
```

trbromley commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
+        postprocessed = qcut.qcut_processing_fn_mc(
+            fixed_samples, communication_graph, fixed_settings, shots, fn
+        )
+        postprocessed = qml.math.cast_like(postprocessed, lib.ones(1))
```
```suggestion
```
I don't think we need this - we should be testing that `postprocessed` is the expected type without casting.

trbromley commented on tests/transforms/test_qcut.py:
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
+        spy_hstack = mocker.spy(np, "hstack")
+
+        postprocessed = qcut.qcut_processing_fn_mc(
+            fixed_samples, communication_graph, fixed_settings, shots, fn
```
```suggestion
            cast_fixed_samples, communication_graph, fixed_settings, shots, fn
```

trbromley reviewed (commented):
Thanks @anthayes92. I've left a final round of comments but should be good to merge once resolved.

anthayes92 commented:
Thanks for the suggestions @trbromley . Updates to the tests have been manually checked and added here: ceb13836b334dbb91fdb3a10f446c2aa2ba38635

anthayes92 commented on pennylane/transforms/qcut.py:
```diff
@@ -755,45 +760,37 @@ def expand_fragment_tapes_mc(
-        1: ─────╰X──┤ Sample[X]
-        M0 =
-        [1]
+        0: ──H─╭C─┤  Sample[|1⟩⟨1|]
```
kudos to the mystery contributor! :eyes:

trbromley reviewed (approved):
Thanks @anthayes92!
